servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/RequestResponseCloseHandler.java

```diff
@@ -236,21 +268,31 @@ void userClosing(final Channel channel) {
 
     // This closes the channel either completely when there are no more outstanding requests to drain or half-closes
     // when a deferred request was encountered.
-    private void closeChannelHalfOrFullyOnPayloadEnd(final Channel channel, final CloseEvent evt) {
+    private void closeChannelHalfOrFullyOnPayloadEnd(final Channel channel, final CloseEvent evt,
+                                                     final boolean endInbound) {
+
         if (idle(pending, state)) {
-            // close when all pending requests drained
-            closeChannel(channel, evt);
-        } else if (!isClient && evt == PROTOCOL_CLOSING_INBOUND) {
-            // deferred half close after current request is done
+            if (isClient || has(state, IN_CLOSED) || (event != USER_CLOSING && event != PROTOCOL_CLOSING_OUTBOUND)) {
```

Comment:
I am concerned about splitting the state b/w `state` and `event`. Can we set `SERVER_CLOSING_GRACEFULLY` upfront so that we do not have to rely on `event` here?

Reply:
`SERVER_CLOSING_GRACEFULLY` used by logic that handles shutdown events and the requirement there is to set this flag only when we shutdown outbound gracefully. I will need another flag if I can not use `evt`.

Here we use the originated reason for closure to decide if we need to close gracefully or not.

Reply:
After investigating it more, I'm actually planning to do the opposite. The `CLOSING` state is not useful in many cases as it does not tell the reason for the closure. I plan to remove `CLOSING` and use `event != null` instead of `has(state, CLOSING)`. In some cases, like this one, the event matters.